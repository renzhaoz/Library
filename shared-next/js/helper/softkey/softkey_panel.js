/* global NavigationMap, LazyLoader, OptionMenu */

/* eslint no-invalid-this: "off" */
const FORM_ID = 'softkeyPanel';
const SoftkeyPanel = function SoftkeyPanel(actions, optionMenuCallback, area) {
  if (!actions || !actions.items || actions.items.length === 0) {
    return;
  }
  this.softKeyArea = area || document.body;
  this.subMenuNeedHide = false;

  this.initSoftKeyPanel = function initSoftKeyPanel(actionlList) {
    if (this.handleDelay) {
      this.pause = true;
    }
    if (actionlList && actionlList.items && actionlList.items.length > 0) {
      if (this.actions && this.actions.items && this.actions.items.length > 0) {
        this.actions.clear();
      }
      this.actions = actionlList.items;
      this.header = actionlList.header || 'header';
      this.handlers.clear();

      /*
       * It is necessary to change softkey bar executing order by restart
       * the listener if the other softkey instance initialized (mainly
       * for CustomDialog)
       */
      this.stopListener();
      initSoftkeys.call(this);
      setSoftkeys.call(this);
      this.startListener();
    }
    if (this.handleDelay) {
      setTimeout(() => {
        this.pause = false;
      }, 200);
    }
  };

  this.startListener = function startListener() {
    window.addEventListener('keydown', this.keyDownHandler, true);
    window.addEventListener('menuChangeEvent', this.keyDownHandler);
  };

  this.stopListener = function stopListener() {
    window.removeEventListener('keydown', this.keyDownHandler, true);
    window.removeEventListener('menuChangeEvent', this.keyDownHandler);
  };

  this.keyDownHandler = function keyDownHandler(e) {
    if (!this.form) {
      return;
    }
    if (!this.form.classList.contains('visible') && !this.menuVisible) {
      return;
    }
    if (
      (e.key === 'Enter' || e.key === 'Accept') &&
      this.menuVisible &&
      this.menu
    ) {
      const focusedButton = this.menu.form.querySelector('.focus');
      let isSubMenu = false;
      if (focusedButton !== null) {
        isSubMenu = focusedButton.dataset.hasmenu || false;
      }
      if (this.menu && !isSubMenu) {
        closeMenu.call(this);
      }
      return;
    }
    if (this.pause) {
      return;
    }
    if (
      typeof NavigationMap !== 'undefined' &&
      NavigationMap.currentActivatedLength > 0
    ) {
      return;
    }
    switch (e.key) {
      case 'BrowserSearch':
      case 'F2': // Device key
      case 'SoftRight': // Emulator key
        if (this.menuVisible || !this.softKeyVisible) {
          break;
        }
        if (this.menuStartIndex === -1) {
          callMethod.call(this, this.buttonRsk);
        } else {
          openMenu.call(this);
        }
        e.preventDefault();
        e.stopPropagation();
        break;
      case 'ContextMenu':
      case 'F1': // Device key
      case 'SoftLeft': // Emulator key
        if (!this.menuVisible && this.softKeyVisible) {
          callMethod.call(this, this.buttonLsk);
          e.preventDefault();
        }
        break;

      /*
       * Case "End": //keybord key
       * case "ContextMenu": //emulator key
       * console.log("RSK relized", this.buttonLsk);
       * callMethod.call(this, this.buttonLsk);
       * break;
       */
      case 'Enter': // Keyboard key
      case 'Accept': // Emulator key
        if (this.menuVisible) {
          closeMenu.call(this);
        } else {
          if (!this.softKeyVisible) {
            break;
          }
          callMethod.call(this, this.buttonCsk);
        }
        break;
      case 'BrowserBack': // Emulator key
      case 'Backspace':
        if (this.menuVisible) {
          e.preventDefault();
        }
        if (this.menu && this.menu.subMenuDisplayed) {
          this.subMenuNeedHide = true;
        }
        closeMenu.call(this);
        break;
      default:
        break;
    }
    if (e.detail !== null) {
      switch (e.detail.action) {
        case 'closeMenu':
          closeMenu.call(this);
          break;
        default:
          break;
      }
    }
  }.bind(this);

  function createForm(formId) {
    const form = document.createElement('form');
    form.id = formId;
    form.dataset.type = this.actions.type || 'action';
    form.className = 'skbar';
    form.classList.add('none-paddings');
    form.setAttribute('data-z-index-level', 'softkeyPanel');
    form.addEventListener('updateKeyInfo', () => {
      modifyAllString.call(this);
    });
    form.addEventListener('submit', event => {
      event.preventDefault();
    });

    // Add/remove
    form.addEventListener('transitionend', event => {
      if (event.target !== form) {
        return;
      }
      if (!form.classList.contains('visible') && form.parentNode) {
        if (this.softKeyArea) {
          this.softKeyArea.removeChild(form);
        } else {
          document.body.removeChild(form);
        }
      }
    });
    return form;
  }

  this.createPanel = function createPanel(form) {
    const lskDiv = document.createElement('div');
    const cskDiv = document.createElement('div');
    const rskDiv = document.createElement('div');

    this.buttonLsk = document.createElement('button');
    this.buttonLsk.className = 'sk-button';
    this.buttonLsk.setAttribute('id', 'software-keys-left');

    this.buttonCsk = document.createElement('button');
    this.buttonCsk.className = 'sk-button';
    this.buttonCsk.setAttribute('id', 'software-keys-center');

    this.buttonRsk = document.createElement('button');
    this.buttonRsk.className = 'sk-button';
    this.buttonRsk.setAttribute('id', 'software-keys-right');

    form.appendChild(lskDiv);
    form.appendChild(cskDiv);
    form.appendChild(rskDiv);

    lskDiv.appendChild(this.buttonLsk);
    cskDiv.appendChild(this.buttonCsk);
    rskDiv.appendChild(this.buttonRsk);

    initSoftkeys.call(this);
    setSoftkeys.call(this);
    this.startListener();
  };

  function initSoftkeys() {
    let lskReady = false;
    let cskReady = false;
    let rskReady = false;
    // Reset indexes
    this.menuStartIndex = -1;
    this.lskIndex = -1;
    this.cskIndex = -1;
    this.rskIndex = -1;

    for (let i = 0; i < this.actions.length; i++) {
      const curPriority = this.actions[i].priority;
      if (curPriority < this.menuActionsPriority) {
        // Trying to initialize softkeys, if they are not initialized yet
        if (!lskReady && curPriority === this.lskPriority) {
          this.lskIndex = i;
          lskReady = true;
        } else if (!cskReady && curPriority === this.cskPriority) {
          this.cskIndex = i;
          cskReady = true;
        } else if (!rskReady) { //eslint-disable-line
          if (i === this.actions.length - 1) {
            this.rskIndex = i;
          } else {
            /*
             * There are > 1 items in the actions list left:
             * - assign 'Options' to RSK and move tail to menu actions.
             */
            this.menuStartIndex = i;
          }
          rskReady = true;
          break;
        } else {
          console.warn('no SK for set actions');
        }
      } else if (!rskReady) {
        /*
         * There are no more actions with sufficient priorities in list
         * - assign 'Options' to RSK and move tail to menu actions.
         */
        this.menuStartIndex = i;
        rskReady = true;
        break;
      }
    }
  }

  function setSoftkeys() {
    function getSkHtml(item) {
      const name = item.l10nId ? window.api.l10n.get(item.l10nId) : item.name;
      const nameWithIcon = item.iconMixedText ? name : '';
      const innerContent = item.iconReverse
        ? `<span data-icon="${item.icon}"></span>
           <span data-name="${name}" class="with-icon-right">${nameWithIcon}</span>`
        : `<span data-name="${name}" class="with-icon-left">${nameWithIcon}</span>
           <span data-icon="${item.icon}"></span>`;
      return item.icon ? innerContent : name;
    }
    let needDisable = false;

    // eslint-disable-next-line
    function updateSoftKeys(actionList, handlers, button, index) {
      const item = actionList[index];
      handlers.set(button, item);
      button.innerHTML = getSkHtml(item);
      if (typeof item.l10nId === 'undefined') {
        if (item.icon) {
          button.firstChild.removeAttribute('data-l10n-id');
        } else {
          button.removeAttribute('data-l10n-id');
        }
      } else if (item.icon) {
        if (item.iconReverse) {
          button.lastChild.setAttribute('data-l10n-id', item.l10nId);
        } else {
          button.firstChild.setAttribute('data-l10n-id', item.l10nId);
        }
      } else {
        button.setAttribute('data-l10n-id', item.l10nId);
      }
      needDisable = item.disabled;
    }

    // Update LSK
    if (this.lskIndex !== -1) { // eslint-disable-line
      updateSoftKeys(
        this.actions,
        this.handlers,
        this.buttonLsk,
        this.lskIndex
      );
    } else {
      this.buttonLsk.innerHTML = '';
      this.buttonLsk.removeAttribute('data-l10n-id');
      needDisable = true;
    }
    disableSk(this.buttonLsk, needDisable);

    // Update CSK
    if (this.cskIndex !== -1) { // eslint-disable-line
      updateSoftKeys(
        this.actions,
        this.handlers,
        this.buttonCsk,
        this.cskIndex
      );
    } else {
      this.buttonCsk.innerHTML = '';
      this.buttonCsk.removeAttribute('data-l10n-id');
      needDisable = true;
    }
    disableSk(this.buttonCsk, needDisable);

    // Update RSK
    if (this.rskIndex !== -1) {
      updateSoftKeys(
        this.actions,
        this.handlers,
        this.buttonRsk,
        this.rskIndex
      );
    } else if (this.menuStartIndex != -1) { // eslint-disable-line
      this.buttonRsk.setAttribute('data-l10n-id', 'softkey-options');
      this.buttonRsk.innerHTML =
        window.api.l10n.get('softkey-options') || 'Options';
      needDisable = false;
    } else {
      this.buttonRsk.innerHTML = '';
      this.buttonRsk.removeAttribute('data-l10n-id');
      needDisable = true;
    }
    disableSk(this.buttonRsk, needDisable);
    modifyAllString.call(this);
  }

  function callMethod(button) {
    const action = this.handlers.get(button);

    if (typeof action !== 'undefined') {
      /* eslint-disable */
      const method = action.method || function () {};

      method.apply(null, action.params || []);

      if (typeof this.actions.complete === 'function' && !action.incomplete) {
        this.actions.complete();
      }
      /* eslint-enable */
    }
  }

  function closeMenu() {
    if (this.menu) {
      const promise = this.menu.hide();
      if (!promise) {
        return;
      }
      promise.then(() => {
        if (this.menu.subMenuDisplayed || this.subMenuNeedHide) {
          this.subMenuNeedHide = false;
          secondCall = false; // eslint-disable-line
        } else {
          this.menuVisible = false;
          setSoftkeys.call(this);
          document.dispatchEvent(this.menuHideEvent);
          window.dispatchEvent(this.menuHideEvent);
        }
      });
    }
  }
  this.hideMenu = function hideMenu() {
    closeMenu.call(this);
  };

  function openMenu() {
    if (document.querySelector('gaia-confirm')) {
      return;
    }
    setOptionMenuSk.call(this);

    let promise = null;
    const omParams = {
      header: this.header,
      items: this.actions
        .slice(this.menuStartIndex)
        .sort((a, b) => a.priority - b.priority),
      classes: ['group-menu', 'softkey'],
      menuClassName: this.menuClassName
    };
    if (typeof OptionMenu === 'undefined') {
      const lazyLoadFiles = [
        '%SHARED_APP_ORIGIN%/style/commons/action_menu.css',
        '%SHARED_APP_ORIGIN%/style/commons/option_menu.css'
      ];
      if (!window.OptionMenu) {
        lazyLoadFiles.push(
          '%SHARED_APP_ORIGIN%/js/helper/option_menu/option_menu.js'
        );
      }
      LazyLoader.load(lazyLoadFiles, () => {
        this.menu = new OptionMenu(omParams, this.optionMenuCallback);
        promise = this.menu.show();
        /* eslint-disable */
        promise && promise.then(() => {
            setOptionMenuSk.call(this);
            modifyAllString.call(this);
            document.dispatchEvent(this.menuShowEvent);
            window.dispatchEvent(this.menuShowEvent);
          });
          /* eslint-enable */
      });
    } else {
      // If(!this.menu)
      secondCall = false; // eslint-disable-line
      this.menu = new OptionMenu(omParams, this.optionMenuCallback);
      promise = this.menu.show();
      /* eslint-disable */
      promise && promise.then(() => { //eslint-disable-line
        setOptionMenuSk.call(this);
        modifyAllString.call(this);
        document.dispatchEvent(this.menuShowEvent);
        window.dispatchEvent(this.menuShowEvent);
      });
      /* eslint-enable */
    }
    this.menuVisible = true;
  }

  function setOptionMenuSk() {
    this.buttonLsk.innerHTML = '';
    disableSk(this.buttonLsk, true);
    this.buttonCsk.innerHTML = window.api.l10n.get('select');
    disableSk(this.buttonCsk, false);
    this.buttonRsk.innerHTML = '';
    disableSk(this.buttonRsk, true);
  }

  function disableSk(skBtn, isDisabled) {
    const alreadyDisabled = skBtn.hasAttribute('disabled');
    if (isDisabled && !alreadyDisabled) {
      skBtn.setAttribute('disabled', 'disabled');
    } else if (!isDisabled && alreadyDisabled) {
      skBtn.removeAttribute('disabled');
    }
  }
  this.handlers = new Map();
  this.menuActionsPriority = 4;
  this.actions = actions.items;
  this.header = actions.header || 'header';
  this.optionMenuCallback = optionMenuCallback;
  this.handleDelay = !!actions.handleDelay;
  this.pause = false;
  this.menuClassName = actions.menuClassName;
  if (typeof this.menuClassName === 'undefined') {
    this.menuClassName = 'menu-button';
  }
  this.menu = null;
  this.menuShowEvent = new CustomEvent('menuEvent', {
    detail: {
      softkeyPanel: this,
      menuName: this.menuClassName,
      menuVisible: true
    }
  });
  this.menuHideEvent = new CustomEvent('menuEvent', {
    detail: {
      softkeyPanel: this,
      menuName: this.menuClassName,
      menuVisible: false
    }
  });
  this.menuVisible = false;
  this.softKeyVisible = false;
  this.lskPriority = 1;
  this.cskPriority = 2;
  this.form = createForm.call(this, FORM_ID);

  this.createPanel(this.form);
  this.getLeftKey = function getLeftKey() {
    return this.buttonLsk;
  };
  this.getRightKey = function getRightKey() {
    return this.buttonRsk;
  };
  this.getCenterKey = function getCenterKey() {
    return this.buttonCsk;
  };
};

SoftkeyPanel.prototype.show = function show() {
  if (!this.form.parentNode) {
    if (this.softKeyArea) {
      this.softKeyArea.appendChild(this.form);
    } else {
      document.body.appendChild(this.form);
    }
  }
  this.form.classList.remove('inactive');
  this.form.classList.add('visible', 'focused');

  this.softKeyVisible = true;
  this.form.focus();
  modifyAllString.call(this);
};

function updateNextSoftkeyPanelKeysInfo() {
  const softkeyPanels = document.querySelectorAll(
    '#softkeyPanel.visible.focused'
  );
  if (softkeyPanels.length) {
    const nextSofkeyPanel = softkeyPanels[softkeyPanels.length - 1];
    nextSofkeyPanel.dispatchEvent(new CustomEvent('updateKeyInfo'));
  }
}

function modifyAllString() {
  const keysInfo = [];

  if (this.buttonLsk.innerHTML.length > 0) {
    // Get string from button before modifyString.
    keysInfo.push(getInfoFromButton(this.buttonLsk, this));
  }

  if (this.buttonCsk.innerHTML.length > 0) {
    // Get string from button before modifyString.
    keysInfo.push(getInfoFromButton(this.buttonCsk, this));
  }

  if (this.buttonRsk.innerHTML.length > 0) {
    // Get string from button before modifyString.
    keysInfo.push(getInfoFromButton(this.buttonRsk, this));
  }

  const softkeyPanels = document.querySelectorAll(
    '#softkeyPanel.visible.focused'
  );
  if (
    softkeyPanels.length &&
    softkeyPanels[softkeyPanels.length - 1] === this.form
  ) {
    if (window.registerSoftkeys) {
      window.registerSoftkeys(keysInfo);
    }
  }
}

function getInfoFromButton(targetElement, self) {
  const info = {};
  switch (targetElement) {
    case self.buttonLsk:
      info.code = 'SoftLeft';
      break;
    case self.buttonCsk:
      info.code = 'Enter';
      break;
    case self.buttonRsk:
      info.code = 'SoftRight';
      break;
    default:
      break;
  }

  const childElement = targetElement.querySelector('span');
  if (childElement) {
    const elementName = childElement.dataset.name;
    info.options = {
      name: elementName ? elementName : '',
      icon: childElement.dataset.icon
    };
  } else {
    info.options = {
      name: targetElement.textContent,
      icon: ''
    };
  }

  return info;
}

// To do Method doesn't work
SoftkeyPanel.prototype.hide = function hide() {
  this.form.classList.remove('visible', 'focused');
  this.softKeyVisible = false;
  if (window.registerSoftkeys) {
    window.registerSoftkeys([]);
  }
  updateNextSoftkeyPanelKeysInfo();
};

SoftkeyPanel.prototype.destroy = function destroy() {
  if (this.softKeyArea) {
    this.softKeyArea.removeChild(this.form);
  } else {
    document.body.removeChild(this.form);
  }
  this.form = null;
  this.stopListener();
  this.softKeyVisible = false;
  this.menuVisible = false;
};
if (!window.registerSoftkeys) {
  LazyLoader.load('%SHARED_APP_ORIGIN%/js/helper/softkey/softkey_register.js');
}

window.SoftkeyPanel = SoftkeyPanel;
