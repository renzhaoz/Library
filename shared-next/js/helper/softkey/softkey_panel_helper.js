/* eslint-disable no-unused-expressions */

(function softkeyPanelHelper(exports) {
  let softkey = null;
  let backCallBack = [];
  let softkey_enabled = false;
  const HelperEvent = function HelperEvent(e) {
    switch (e.type) {
      case 'screenchange':
        if (!e.detail.screenEnabled) {
          softkey && softkey.hide();
        }
        break;
      case 'lockscreen-appopened':
        softkey && softkey.hide();
        break;
      case 'lockscreen-appclosed':
      case 'wake':
        if (softkey_enabled) {
          softkey && softkey.show();
        }
        break;
      case 'keydown':
        if (e.key === 'Backspace' || e.key === 'BrowserBack') {
          if (softkey_enabled) {
            // eslint-disable-next-line no-useless-call
            backCallBack.apply(null, []);
            e.preventDefault();
          }
        }
        break;
      default:
        break;
    }
  };

  window.addEventListener('keydown', HelperEvent);
  window.addEventListener('lockscreen-appopened', HelperEvent);
  window.addEventListener('lockscreen-appclosed', HelperEvent);
  window.addEventListener('screenchange', HelperEvent);
  window.addEventListener('wake', HelperEvent);

  const SoftkeyHelper = {
    applist: [],
    init(params, callback) {
      if (!softkey) {
        // eslint-disable-next-line no-undef
        softkey = new SoftkeyPanel({
          menuClassName: 'menu-button',
          items: [
            {
              name: 'create',
              priority: 1,
              // eslint-disable-next-line no-empty-function
              method() {}
            }
          ]
        });
      }
      softkey.initSoftKeyPanel(params);
      softkey.show();
      // eslint-disable-next-line no-empty-function
      backCallBack = callback || function nc() {};
      softkey_enabled = true;
    },
    updateSoftkey(params, keyStyle, name) {
      switch (keyStyle) {
        case '1':
          params.items[0].l10nId = name;
          break;
        case '3':
          params.items[1].l10nId = name;
          break;
        default:
          break;
      }
      softkey && softkey.initSoftKeyPanel(params);
    },

    show() {
      softkey && softkey.show();
    },

    onlyHide() {
      softkey && softkey.hide();
    },

    hide(name) {
      softkey && softkey.hide();
      // eslint-disable-next-line no-empty-function,func-names
      backCallBack = function() {};
      softkey_enabled = false;
      for (let i = 0; i < this.applist.length; i++) {
        if (this.applist[i].name === name) {
          this.applist.splice(i, 1);
          if (
            i === 0 &&
            this.applist.length > 0 &&
            this.applist[0].appliacation.front !== undefined
          ) {
            this.applist[0].appliacation.front();
          }
          break;
        }
      }
    },
    getSoftkey() {
      return softkey;
    },
    register(app, name) {
      for (let i = 0; i < this.applist.length; i++) {
        if (this.applist[i].name === name) {
          return;
        }
      }
      const applicationData = {
        appliacation: app,
        name
      };
      if (
        this.applist.length > 0 &&
        this.applist[0].appliacation.back !== undefined
      ) {
        this.applist[0].appliacation.back();
      }
      this.applist.unshift(applicationData);
    }
  };

  exports.SoftkeyHelper = SoftkeyHelper;
  // eslint-disable-next-line no-invalid-this
})(window);
