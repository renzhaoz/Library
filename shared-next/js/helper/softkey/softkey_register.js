/*
 * This function is used for send softkey content to system app for readout
 * keysInfo format:[
 *   { 'code': 'SoftLeft', // 'SoftLeft', 'Enter', 'SoftRight',
 *     'name': 'cancle'
 *   },
 *   {...},
 *   ...
 *  ]
 */
function registerSoftkeys(keysInfo) {
  // RegisterKeys via softkeyManager
  if (keysInfo) {
    const metaName = 'og:kaios:softkeyinfo';
    const softKeys = [];
    let softkeyInfoMeta = document.head.querySelector(
      `meta[name="${metaName}"]`
    );
    if (!softkeyInfoMeta) {
      softkeyInfoMeta = document.createElement('meta');
      softkeyInfoMeta.name = metaName;
      document.head.appendChild(softkeyInfoMeta);
    }
    keysInfo.forEach(key => {
      if (key.options.name) {
        if (typeof key.options.name === 'object') {
          if (key.options.name.text) {
            softKeys.push(`${key.code}: ${key.options.name.text}`);
          }
        } else {
          softKeys.push(`${key.code}: ${key.options.name}`);
        }
      }
    });
    softkeyInfoMeta.setAttribute('content', softKeys.join(', '));
  }
}

window.registerSoftkeys = registerSoftkeys;
