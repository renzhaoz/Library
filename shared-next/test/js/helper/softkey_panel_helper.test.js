/* eslint-disable no-undef, global-require */
describe('helper js <softkey_panel_helper> test', () => {
  const { wrap } = require('jest-snapshot-serializer-raw');
  const back = jest.fn();
  const front = jest.fn();
  const app = { back, front };
  const callback = jest.fn();
  let softkey = null;

  beforeAll(done => {
    require('../../mock/l10n_mock');
    require('../../mock/lazyLoader_mock');
    require('../../mock/softkey_register_mock');
    require('../../../js/helper/softkey/softkey_panel');
    require('../../../js/helper/softkey/softkey_panel_helper');
    done();
  });

  // Test SoftkeyHelper
  test('SoftkeyHelper should be object', done => {
    expect(typeof window.SoftkeyHelper).toBe('object');
    done();
  });

  // Test SoftkeyHelper.init
  test('SoftkeyHelper.init should be function', done => {
    expect(typeof window.SoftkeyHelper.init).toBe('function');
    done();
  });

  // Test SoftkeyHelper.updateSoftkey
  test('SoftkeyHelper.updateSoftkey should be function', done => {
    expect(typeof window.SoftkeyHelper.updateSoftkey).toBe('function');
    done();
  });

  // Test SoftkeyHelper.show
  test('SoftkeyHelper.show should be function', done => {
    expect(typeof window.SoftkeyHelper.show).toBe('function');
    done();
  });

  // Test SoftkeyHelper.onlyHide
  test('SoftkeyHelper.onlyHide should be function', done => {
    expect(typeof window.SoftkeyHelper.onlyHide).toBe('function');
    done();
  });

  // Test SoftkeyHelper.hide
  test('SoftkeyHelper.hide should be function', done => {
    expect(typeof window.SoftkeyHelper.hide).toBe('function');
    done();
  });

  // Test SoftkeyHelper.getSoftkey
  test('SoftkeyHelper.getSoftkey should be function', done => {
    expect(typeof window.SoftkeyHelper.getSoftkey).toBe('function');
    done();
  });

  // Test SoftkeyHelper.register
  test('SoftkeyHelper.register should be function', done => {
    expect(typeof window.SoftkeyHelper.register).toBe('function');
    done();
  });

  // Test SoftkeyHelper.register return value
  test('SoftkeyHelper.register should return applist', done => {
    const name = 'test';
    SoftkeyHelper.register(app, name);
    expect(SoftkeyHelper.applist.length).toBe(1);
    SoftkeyHelper.register({}, name);
    expect(SoftkeyHelper.applist.length).toBe(1);
    SoftkeyHelper.register(app, 'test1');
    expect(SoftkeyHelper.applist.length).toBe(2);
    expect(back).toBeCalledTimes(1);
    done();
  });

  describe('SoftkeyHelper.init', () => {
    beforeEach(done => {
      const softkeyParams = {
        menuClassName: 'menu-button',
        header: { name: 'Ring' },
        items: [
          {
            l10nId: 'ringer-dismiss',
            priority: 3,
            method: jest.fn()
          }
        ]
      };
      SoftkeyHelper.init(softkeyParams, callback);
      softkey = SoftkeyHelper.getSoftkey();
      done();
    });

    // Test SoftkeyHelper.init
    test('SoftkeyHelper.init should create soft key panel', done => {
      expect(wrap(softkey.form)).toMatchSnapshot();
      expect(wrap(softkey.buttonLsk)).toMatchSnapshot();
      expect(wrap(softkey.buttonCsk)).toMatchSnapshot();
      expect(wrap(softkey.buttonRsk)).toMatchSnapshot();
      expect(typeof softkey.show).toBe('function');
      expect(typeof softkey.hide).toBe('function');
      expect(softkey.form.classList.contains('visible')).toBeTruthy();
      expect(softkey.form.classList.contains('focused')).toBeTruthy();
      done();
    });

    // Test HelperEvent
    test('HelperEvent should show/hide softkey form', done => {
      window.dispatchEvent(
        new KeyboardEvent('keydown', { type: 'keydown', key: 'Backspace' })
      );
      window.dispatchEvent(
        new KeyboardEvent('keydown', { type: 'keydown', key: 'BrowserBack' })
      );
      expect(callback).toBeCalledTimes(2);
      window.dispatchEvent(
        new CustomEvent('lockscreen-appopened', { type: 'lockscreen-appopened' })
      );
      expect(softkey.form.classList.contains('visible')).toBeFalsy();
      expect(softkey.form.classList.contains('focused')).toBeFalsy();
      window.dispatchEvent(
        new CustomEvent('lockscreen-appclosed', { type: 'lockscreen-appclosed' })
      );
      expect(softkey.form.classList.contains('visible')).toBeTruthy();
      expect(softkey.form.classList.contains('focused')).toBeTruthy();
      window.dispatchEvent(
        new CustomEvent('screenchange', {
          type: 'screenchange',
          detail: { screenEnabled: false }
        })
      );
      expect(softkey.form.classList.contains('visible')).toBeFalsy();
      expect(softkey.form.classList.contains('focused')).toBeFalsy();
      done();
    });

    // Test SoftkeyHelper.show
    test('SoftkeyHelper.show should show form', done => {
      window.SoftkeyHelper.show();
      expect(softkey.form.classList.contains('visible')).toBeTruthy();
      expect(softkey.form.classList.contains('focused')).toBeTruthy();
      done();
    });

    // Test SoftkeyHelper.onlyHide
    test('SoftkeyHelper.onlyHide should hide form', done => {
      window.SoftkeyHelper.onlyHide();
      expect(softkey.form.classList.contains('visible')).toBeFalsy();
      expect(softkey.form.classList.contains('focused')).toBeFalsy();
      done();
    });

    // Test SoftkeyHelper.hide
    test('SoftkeyHelper.hide should hide form', done => {
      SoftkeyHelper.register(app, 'test');
      SoftkeyHelper.register(app, 'test1');
      expect(window.SoftkeyHelper.applist.length).toBe(2);
      window.SoftkeyHelper.hide('test1');
      expect(softkey.form.classList.contains('visible')).toBeFalsy();
      expect(softkey.form.classList.contains('focused')).toBeFalsy();
      expect(window.SoftkeyHelper.applist.length).toBe(1);
      expect(front).toBeCalledTimes(1);
      done();
    });

    // Test SoftkeyHelper.updateSoftkey
    test('SoftkeyHelper.updateSoftkey should update Softkey panel', done => {
      const params = {
        items: [
          {
            l10nId: 'ringer-dismiss',
            priority: 2,
            method: jest.fn()
          },
          {
            l10nId: 'ringer-dismiss',
            priority: 3,
            method: jest.fn()
          }
        ]
      };
      window.SoftkeyHelper.updateSoftkey(params, '1', 'new-ringer-dismiss1');
      window.SoftkeyHelper.updateSoftkey(params, '2', 'new-ringer-dismiss2');
      window.SoftkeyHelper.updateSoftkey(params, '3', 'new-ringer-dismiss3');
      expect(wrap(softkey.form)).toMatchSnapshot();
      expect(wrap(softkey.buttonLsk)).toMatchSnapshot();
      expect(wrap(softkey.buttonCsk)).toMatchSnapshot();
      expect(wrap(softkey.buttonRsk)).toMatchSnapshot();
      done();
    });
  });
});
