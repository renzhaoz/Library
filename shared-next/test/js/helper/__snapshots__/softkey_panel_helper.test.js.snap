// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helper js <softkey_panel_helper> test SoftkeyHelper.init SoftkeyHelper.init should create soft key panel 1`] = `
Object {
  Symbol(jest-snapshot-serializer-raw): <form
    class="skbar none-paddings visible focused"
    data-type="action"
    data-z-index-level="softkeyPanel"
    id="softkeyPanel"
  >
    <div>
      <button
        class="sk-button"
        disabled="disabled"
        id="software-keys-left"
      />
    </div>
    <div>
      <button
        class="sk-button"
        disabled="disabled"
        id="software-keys-center"
      />
    </div>
    <div>
      <button
        class="sk-button"
        data-l10n-id="ringer-dismiss"
        id="software-keys-right"
      >
        ringer-dismiss
      </button>
    </div>
  </form>,
}
`;

exports[`helper js <softkey_panel_helper> test SoftkeyHelper.init SoftkeyHelper.init should create soft key panel 2`] = `
Object {
  Symbol(jest-snapshot-serializer-raw): <button
    class="sk-button"
    disabled="disabled"
    id="software-keys-left"
  />,
}
`;

exports[`helper js <softkey_panel_helper> test SoftkeyHelper.init SoftkeyHelper.init should create soft key panel 3`] = `
Object {
  Symbol(jest-snapshot-serializer-raw): <button
    class="sk-button"
    disabled="disabled"
    id="software-keys-center"
  />,
}
`;

exports[`helper js <softkey_panel_helper> test SoftkeyHelper.init SoftkeyHelper.init should create soft key panel 4`] = `
Object {
  Symbol(jest-snapshot-serializer-raw): <button
    class="sk-button"
    data-l10n-id="ringer-dismiss"
    id="software-keys-right"
  >
    ringer-dismiss
  </button>,
}
`;

exports[`helper js <softkey_panel_helper> test SoftkeyHelper.init SoftkeyHelper.updateSoftkey should update Softkey panel 1`] = `
Object {
  Symbol(jest-snapshot-serializer-raw): <form
    class="skbar none-paddings visible focused"
    data-type="action"
    data-z-index-level="softkeyPanel"
    id="softkeyPanel"
  >
    <div>
      <button
        class="sk-button"
        disabled="disabled"
        id="software-keys-left"
      />
    </div>
    <div>
      <button
        class="sk-button"
        data-l10n-id="new-ringer-dismiss1"
        id="software-keys-center"
      >
        new-ringer-dismiss1
      </button>
    </div>
    <div>
      <button
        class="sk-button"
        data-l10n-id="new-ringer-dismiss3"
        id="software-keys-right"
      >
        new-ringer-dismiss3
      </button>
    </div>
  </form>,
}
`;

exports[`helper js <softkey_panel_helper> test SoftkeyHelper.init SoftkeyHelper.updateSoftkey should update Softkey panel 2`] = `
Object {
  Symbol(jest-snapshot-serializer-raw): <button
    class="sk-button"
    disabled="disabled"
    id="software-keys-left"
  />,
}
`;

exports[`helper js <softkey_panel_helper> test SoftkeyHelper.init SoftkeyHelper.updateSoftkey should update Softkey panel 3`] = `
Object {
  Symbol(jest-snapshot-serializer-raw): <button
    class="sk-button"
    data-l10n-id="new-ringer-dismiss1"
    id="software-keys-center"
  >
    new-ringer-dismiss1
  </button>,
}
`;

exports[`helper js <softkey_panel_helper> test SoftkeyHelper.init SoftkeyHelper.updateSoftkey should update Softkey panel 4`] = `
Object {
  Symbol(jest-snapshot-serializer-raw): <button
    class="sk-button"
    data-l10n-id="new-ringer-dismiss3"
    id="software-keys-right"
  >
    new-ringer-dismiss3
  </button>,
}
`;
